x-pack/solutions/observability/plugins/apm/public/components/shared/trace_waterfall/trace_waterfall_context.tsx

```diff
@@ -162,7 +176,29 @@ export function TraceWaterfallContextProvider({
   const left = TOGGLE_BUTTON_WIDTH + ACCORDION_PADDING_LEFT * maxDepth;
   const right = 40;
 
-  const traceWaterfallMap = useMemo(() => groupByParent(traceWaterfall), [traceWaterfall]);
+  const fullTraceWaterfallMap = useMemo(() => groupByParent(traceWaterfall), [traceWaterfall]);
+
+  const criticalPathSegmentsById = useMemo(() => {
+    if (!showCriticalPath) {
+      return {};
+    }
+
+    const root = traceWaterfall[0];
+    if (!root) {
+      return {};
+    }
+
+    const criticalPath = getCriticalPath(root, fullTraceWaterfallMap);
+    return groupBy(criticalPath.segments, (segment) => segment.item.id);
```

Comment:
Couldn't this just be a reduce on a object? I'm a bit hesitant to allow lodash on new code unless we are doing something that can't be expressed easily with native JS code/APIs.